Issue58: Notifications: Step 2: GraphQL change

ghormadesoham opened the pull request:
# Changes

- Add UserPersonalizationInputType for mutation and query. 
- Declare the query, update and upsert operations.

# Testing
Verify that the website builds and loads with this change.

# Related
- #1 Database change 
- #4 server side change
- Pending : client side change

ghormadesoham committed:
Issue58: GraphQL: Add UserPersonalizationInputType for mutation and query. Declare the query, update and upsert operations.
Testing:Verify that the website builds and loads with this change.

ghormadesoham commented:
# Developer validation testing 

## npm run generate output 
This is the output after generating graphql artifacts:
```
(base) sohamghormade@Sohams-MacBook-Air graphql % npm run generate

> omscentral-graphql@0.0.0 generate
> VERBOSE=true graphql-codegen --config codegen.yml && cp ./schema.graphql ../server/src/graphql/schema.graphql

[07:39:41] Parse configuration [started]
[07:39:41] Parse configuration [completed]
[07:39:41] Generate outputs [started]
[07:39:41] Generate ../server/src/graphql/graphql.schema.json [started]
[07:39:41] Generate ../server/src/graphql/index.ts [started]
[07:39:41] Generate ../client/src/graphql/graphql.schema.json [started]
[07:39:41] Generate ../client/src/graphql/index.tsx [started]
[07:39:41] Load GraphQL schemas [started]
[07:39:41] Load GraphQL schemas [started]
[07:39:41] Load GraphQL schemas [started]
[07:39:41] Load GraphQL schemas [started]
[07:39:41] Load GraphQL schemas [completed]
[07:39:41] Load GraphQL documents [started]
[07:39:41] Load GraphQL schemas [completed]
[07:39:41] Load GraphQL documents [started]
[07:39:41] Load GraphQL schemas [completed]
[07:39:41] Load GraphQL documents [started]
[07:39:41] Load GraphQL schemas [completed]
[07:39:41] Load GraphQL documents [started]
[07:39:41] Load GraphQL documents [completed]
[07:39:41] Generate [started]
[07:39:41] Generate [completed]
[07:39:41] Generate ../server/src/graphql/index.ts [completed]
[07:39:41] Load GraphQL documents [completed]
[07:39:41] Generate [started]
[07:39:42] Generate [completed]
[07:39:42] Generate ../client/src/graphql/index.tsx [completed]
[07:39:42] Load GraphQL documents [completed]
[07:39:42] Generate [started]
[07:39:42] Generate [completed]
[07:39:42] Generate ../server/src/graphql/graphql.schema.json [completed]
[07:39:42] Load GraphQL documents [completed]
[07:39:42] Generate [started]
[07:39:42] Generate [completed]
[07:39:42] Generate ../client/src/graphql/graphql.schema.json [completed]
[07:39:42] Generate outputs [completed]
(base) sohamghormade@Sohams-MacBook-Air graphql
```

ghormadesoham committed:
Issue58: graphql:Fix issue of upsert and update mutations not being identified. The fix was to include them in the schema.graphql since they were missed in the previous commit.
Testing: verify that the change builds and website pages load locally without issue.

mehmetbajin commented on graphql/schema.graphql:
```diff
@@ -192,6 +192,20 @@ type User {
+  "The id corresponding to the user."
+  user_id: String!
+  "Type of personalization e.g frequency"
+  type: String
```
Can we make this a GraphQL `enum` instead?

mehmetbajin commented on graphql/schema.graphql:
```diff
@@ -259,6 +280,13 @@ input UserInputType {
 
+input UserPersonalizationInputType {
+  id: String!
+  user_id: String!
```
We shouldn't need to pass `user_id` in from the client explicitly. Instead, we should use the ID of the currently authenticated user. This way, a given user can only mutate that user's personalizations.

ghormadesoham replied:
Good point. 
In fact, ```UserPersonalizationInputType``` can removed entirely and {id: id, user_id:user.uid, type:'frequency', value:'daily'} can be passed to upsertUserPersonalization mutation.
I will push this change in my next commit.

mehmetbajin commented on graphql/schema.graphql:
```diff
@@ -277,6 +305,16 @@ type Mutation {
+  ): UserPersonalization!
+
+  "Updates a user personalization record."
+  updateUserPersonalization(
+    userPersonalization: UserPersonalizationInputType!
+  ): UserPersonalization!
```
Is `update` needed if we already have `upsert`?

ghormadesoham replied:
Upsert was added for user_personalization because User had both upsert and update.
Notably, in this case, given that the upsert calls update, update will be needed.
Hope this provides context as to why update is still needed.
Thanks.

mehmetbajin reviewed (changes requested):
Couple questions.

ghormadesoham committed:
GraphQL:
1. Remove user_id as a field from UserPersonalizationInputType since it is not required.
2. Propagate DB change from DB PR: Make type and value NotNullable. Use ! in the declaration since exclamation mark indicates that a given field is non nullable in TypeScript.

Pending: client:Create a new UUID and use it as id everytime UserPersonalizationInputType is used.
